sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -485,6 +509,105 @@ case class GpuCast(
+    // The brackets that are used in casting structs and maps to strings
+    val (leftBracket, rightBracket) = if (legacyCastToString) ("[", "]") else ("{", "}")
+    var separatorColumn: ColumnVector = null
+    var spaceColumn: ColumnVector = null
+    val columns: ArrayBuffer[ColumnVector] = new ArrayBuffer[ColumnVector]()
+    // coreColumns tracks the casted child columns
+    val coreColumns: ArrayBuffer[ColumnVector] = new ArrayBuffer[ColumnVector]()
+
+    try {
+      withResource(GpuScalar.from(leftBracket, StringType)) { bracketScalar =>
+        columns += ColumnVector.fromScalar(bracketScalar, input.getRowCount().toInt)
+      }
+      withResource(input.getBase().getChildColumnView(0)) { childView =>
+        withResource(childView.copyToColumnVector()) { childVector =>
+          columns += doColumnar(GpuColumnVector.from(childVector, inputSchema(0).dataType))
+        }
+      }
+
+      if(legacyCastToString) {
+        coreColumns += columns.last
+        withResource(GpuScalar.from(",", StringType)) { separatorScalar =>
+          separatorColumn = ColumnVector.fromScalar(separatorScalar, input.getRowCount().toInt)
+        }
+        withResource(GpuScalar.from(" ", StringType)) { separatorScalar =>
+          spaceColumn = ColumnVector.fromScalar(separatorScalar, input.getRowCount().toInt)
+        }
+        for(childIndex <- 1 until input.getBase().getNumChildren()) {
+          withResource(input.getBase().getChildColumnView(childIndex)) { childView =>
+            columns += separatorColumn
+            // Copies the whitespace column's validity with the current column's validity.
+            // Mimics the Spark null behavior of consecutive commas with no space between them
+            columns += spaceColumn.mergeAndSetValidity(BinaryOp.BITWISE_AND, childView)
+            withResource(childView.copyToColumnVector()) { childVector =>
+              columns += doColumnar(GpuColumnVector.from(childVector,
+                inputSchema(childIndex).dataType))
+              coreColumns += columns.last
+            }
+          }
+        }
+        withResource(GpuScalar.from(rightBracket, StringType)) { bracketScalar =>
+          columns += ColumnVector.fromScalar(bracketScalar, input.getRowCount().toInt)
+        }
+
+        // Merge casted child columns
+        withResource(GpuScalar.from("", StringType)) { emptyStrScalar =>
+          withResource(ColumnVector.stringConcatenate(emptyStrScalar, emptyStrScalar,
+            columns.toArray[ColumnView])) { fullResult =>
+            // Merge the validity of all child columns, fully null rows are null in the result
+            withResource(fullResult.mergeAndSetValidity(BinaryOp.BITWISE_OR,
+              coreColumns: _*)) { nulledResult =>
+              // Reflect the struct column's validity vector in the result
+              nulledResult.mergeAndSetValidity(BinaryOp.BITWISE_AND, input.getBase(), nulledResult)
+            }
+          }
+        }
+      } else {
```

Comment:
The code in the if and else are a little long. it might be nice to either move it into separate methods, or find a way to commonize the code a bit more.

Reply:
They approaches are different enough that I opted to split the code into separate methods. There's a lot of similarities, but how nulls are handled is too different to cleanly parameterize.